haskell_cabal_library linker errors when using GHC 8.10

**Describe the bug**
Building even a simple `haskell_cabal_library` fails with linker errors when using GHC 8.10. The same code built without Cabal does not lead to linker errors.

**To Reproduce**
Switch `rules_haskell` to GHC 8.10.1 or 8.10.2. Then,

```
$ bazel build @stackage//:void
[undefined reference to various RTS symbols errors]
```

**Expected behavior**
I expect the build to succeed with no linker errors.

**Environment**
 - OS name + version: NixOS 20.03 (using a Nixpkgs toolchain) or Debian buster (using GHC bindists).

**Additional context**
This bug has already been reported multiple times: see #1401 #1413 #1327. I'm creating a new issue to gather what we know in a single place, with a more focused issue description.

According to https://github.com/tweag/rules_haskell/issues/1327#issuecomment-664378765, this is a regression introduced by #1388, which is itself necessary on some systems when using GHC 8.8. Since we don't use GHC 8.10 during CI testing, we didn't catch this.

On a PIE-by-default systems like Debian Buster,
* building a `haskell_library` using GHC 8.8 works fine, with or without #1388,
* building a `haskell_cabal_library` using GHC 8.8 works fine, with or without #1388,
* building a `haskell_library` using GHC 8.10 works fine, with or without #1388,
* building a `haskell_cabal_library` using GHC 8.10 works but only without #1388.

The latter case also fails on NixOS with a Nixpkgs toolchain.

So what's still unclear is why the result matrix for `haskell_cabal_library` is different on GHC 8.8 and on GHC 8.10. It's likely that https://gitlab.haskell.org/ghc/ghc/-/issues/15319 has something to do with it.

What's also unclear is why for GHC 8.10, the result matrix is different for `haskell_library` and `haskell_cabal_library`. The former works after #1388 while the latter does not.

cc @patrickt @robinbb

--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -213,9 +213,12 @@ def _prepare_cabal_inputs(
         for arg in ["-package-db", "./" + _dirname(package_db)]
     ], join_with = " ", format_each = "--ghc-arg=%s", omit_if_empty = False)
     args.add("--flags=" + " ".join(flags))
-    if dynamic_file and not (is_library or hs.toolchain.is_darwin or hs.toolchain.is_windows):
+    if dynamic_file:
         # See Note [No PIE when linking] in haskell/private/actions/link.bzl
-        args.add("--ghc-option=-optl-no-pie")
+        if not (hs.toolchain.is_darwin or hs.toolchain.is_windows):
+            version = [int(x) for x in hs.toolchain.version.split(".")]
+            if version < [8, 10] or not is_library:
+                args.add("--ghc-option=-optl-no-pie")
     args.add_all(hs.toolchain.cabalopts)
     args.add_all(cabalopts)
     if dynamic_file:

--- a/haskell/cabal.bzl
+++ b/haskell/cabal.bzl
@@ -140,7 +140,8 @@ def _prepare_cabal_inputs(
         package_database,
         verbose,
         transitive_haddocks,
-        dynamic_binary = None):
+        is_library = False,
+        dynamic_file = None):
     """Compute Cabal wrapper, arguments, inputs."""
     with_profiling = is_profiling_enabled(hs)
 
@@ -212,16 +213,16 @@ def _prepare_cabal_inputs(
         for arg in ["-package-db", "./" + _dirname(package_db)]
     ], join_with = " ", format_each = "--ghc-arg=%s", omit_if_empty = False)
     args.add("--flags=" + " ".join(flags))
-    if not hs.toolchain.is_darwin and not hs.toolchain.is_windows:
+    if dynamic_file and not (is_library or hs.toolchain.is_darwin or hs.toolchain.is_windows):
         # See Note [No PIE when linking] in haskell/private/actions/link.bzl
         args.add("--ghc-option=-optl-no-pie")
     args.add_all(hs.toolchain.cabalopts)
     args.add_all(cabalopts)
-    if dynamic_binary:
+    if dynamic_file:
         args.add_all(
             [
                 "--ghc-option=-optl-Wl,-rpath," + create_rpath_entry(
-                    binary = dynamic_binary,
+                    binary = dynamic_file,
                     dependency = lib,
                     keep_filename = False,
                     prefix = relative_rpath_prefix(hs.toolchain.is_darwin),
@@ -426,7 +427,8 @@ def _haskell_cabal_library_impl(ctx):
         cabal_wrapper = ctx.executable._cabal_wrapper,
         package_database = package_database,
         verbose = ctx.attr.verbose,
-        dynamic_binary = dynamic_library,
+        is_library = True,
+        dynamic_file = dynamic_library,
         transitive_haddocks = _gather_transitive_haddocks(ctx.attr.deps),
     )
     outputs = [
@@ -702,7 +704,7 @@ def _haskell_cabal_binary_impl(ctx):
         cabal_wrapper = ctx.executable._cabal_wrapper,
         package_database = package_database,
         verbose = ctx.attr.verbose,
-        dynamic_binary = binary,
+        dynamic_file = binary,
         transitive_haddocks = _gather_transitive_haddocks(ctx.attr.deps),
     )
     ctx.actions.run(

--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -476,6 +476,12 @@ def compile_library(
     if with_shared:
         c.args.add("-dynamic-too")
 
+        # See Note [No PIE when linking] in haskell/private/actions/link.bzl
+        if not hs.toolchain.is_darwin and not hs.toolchain.is_windows:
+            version = [int(x) for x in hs.toolchain.version.split(".")]
+            if version < [8, 10]:
+                c.args.add("-optl-no-pie")
+
     coverage_data = []
     if hs.coverage_enabled:
         c.args.add_all(_hpc_compiler_args(hs))

--- a/haskell/private/actions/compile.bzl
+++ b/haskell/private/actions/compile.bzl
@@ -475,9 +475,6 @@ def compile_library(
     c = _compilation_defaults(hs, cc, java, posix, dep_info, plugin_dep_info, srcs, import_dir_map, extra_srcs, user_compile_flags, with_profiling, my_pkg_id = my_pkg_id, version = my_pkg_id.version, plugins = plugins, preprocessors = preprocessors)
     if with_shared:
         c.args.add("-dynamic-too")
-        if not hs.toolchain.is_darwin and not hs.toolchain.is_windows:
-            # See Note [No PIE when linking] in haskell/private/actions/link.bzl
-            c.args.add("-optl-no-pie")
 
     coverage_data = []
     if hs.coverage_enabled:

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -213,7 +213,7 @@ def link_binary(
 # on this setting GHC would then automatically set `-no-pie`. However,
 # rules_haskell uses GHC's `-pgmc` flag to point GHC to the CC toolchain's C
 # compiler. This disables the flags configured in the `setting` file. Instead,
-# we have to pass `-no-pie` explicitly.
+# we have to pass `-no-pie` explicitly when linking binaries.
 #
 # Ideally, we would determine whether the CC toolchain's C compiler supports
 # `-no-pie` before setting it. Unfortunately, this is complicated by the fact

--- a/haskell/private/actions/link.bzl
+++ b/haskell/private/actions/link.bzl
@@ -215,6 +215,9 @@ def link_binary(
 # compiler. This disables the flags configured in the `setting` file. Instead,
 # we have to pass `-no-pie` explicitly when linking binaries.
 #
+# In GHC < 8.10, we must always pass -no-pie. In newer compilers, we
+# must take care to only pass -no-pie when compiling libraries.
+#
 # Ideally, we would determine whether the CC toolchain's C compiler supports
 # `-no-pie` before setting it. Unfortunately, this is complicated by the fact
 # that Bazel does not support dynamic dependencies in build actions and that

